File: app/src/main/java/jp/hazuki/yuzubrowser/utils/view/swipebutton/SwipeImageButton.java
package jp.hazuki.yuzubrowser.utils.view.swipebutton;

import android.content.Context;
import android.support.v7.widget.AppCompatImageButton;
import android.util.AttributeSet;
import android.view.MotionEvent;

import jp.hazuki.yuzubrowser.R;
import jp.hazuki.yuzubrowser.action.ActionCallback;
import jp.hazuki.yuzubrowser.action.manager.SoftButtonActionFile;
import jp.hazuki.yuzubrowser.settings.data.ThemeData;

public class SwipeImageButton extends AppCompatImageButton implements SwipeController.OnChangeListener {
    private final SwipeSoftButtonController mController;

    public SwipeImageButton(Context context) {
        this(context, null);
    }

    public SwipeImageButton(Context context, AttributeSet attrs) {
        super(context, attrs);
        mController = new SwipeSoftButtonController(getContext().getApplicationContext());
        setScaleType(ScaleType.CENTER_INSIDE);
    }

    public void setActionData(SoftButtonActionFile action_list, ActionCallback callback) {
        mController.setActionData(action_list, callback);
        mController.setOnChangeListener(this);
        setImageDrawable(mController.getDefaultIcon());
    }

    public void notifyChangeState() {
        mController.notifyChangeState();
    }

    public void setSense(int sense) {
        mController.setSense(sense);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        mController.onTouchEvent(event);
        return super.onTouchEvent(event);
    }

    @Override
    public void onLongPress() {
    }

    @Override
    public boolean onEventOutSide() {
        setImageDrawable(mController.getDefaultIcon());
        setBackgroundResource(R.drawable.swipebtn_image_background_normal);
        return false;
    }

    @Override
    public boolean onEventCancel() {
        setImageDrawable(mController.getDefaultIcon());
        setBackgroundResource(R.drawable.swipebtn_image_background_normal);
        return false;
    }

    @Override
    public boolean onEventActionUp(int whatNo) {
        setImageDrawable(mController.getDefaultIcon());
        setBackgroundResource(R.drawable.swipebtn_image_background_normal);
        return false;
    }

    @Override
    public boolean onEventActionDown() {
        if (ThemeData.isEnabled() && ThemeData.getInstance().toolbarButtonBackgroundPress != null)
            setBackground(ThemeData.getInstance().toolbarButtonBackgroundPress);
        else
            setBackgroundResource(R.drawable.swipebtn_image_background_pressed);
        return false;
    }

    @Override
    public void onChangeState(int whatNo) {
        setImageDrawable(mController.getIcon());
    }
}

File: app/src/main/java/jp/hazuki/yuzubrowser/utils/view/swipebutton/SwipeImageButton.kt
/*
 * Copyright (C) 2017 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.utils.view.swipebutton

import android.content.Context
import android.support.v7.widget.AppCompatImageButton
import android.util.AttributeSet
import android.view.MotionEvent

import jp.hazuki.yuzubrowser.R
import jp.hazuki.yuzubrowser.action.manager.ActionController
import jp.hazuki.yuzubrowser.action.manager.ActionIconManager
import jp.hazuki.yuzubrowser.action.manager.SoftButtonActionFile
import jp.hazuki.yuzubrowser.theme.ThemeData

class SwipeImageButton @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : AppCompatImageButton(context, attrs), SwipeController.OnChangeListener {
    private val mController = SwipeSoftButtonController(getContext().applicationContext)

    init {
        scaleType = ScaleType.CENTER_INSIDE
    }

    fun setActionData(action_list: SoftButtonActionFile, controller: ActionController, iconManager: ActionIconManager) {
        mController.setActionData(action_list, controller, iconManager)
        mController.setOnChangeListener(this)
        setImageDrawable(mController.defaultIcon)
        setBackgroundResource(R.drawable.swipebtn_image_background_normal)
    }

    fun setToDefault() {
        mController.setToDefault()
    }

    fun notifyChangeState() {
        mController.notifyChangeState()
    }

    fun setSense(sense: Int) {
        mController.setSense(sense)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        mController.onTouchEvent(event)
        return super.onTouchEvent(event)
    }

    override fun onLongPress() {}

    override fun onEventOutSide(): Boolean {
        setImageDrawable(mController.defaultIcon)
        setBackgroundResource(R.drawable.swipebtn_image_background_normal)
        return false
    }

    override fun onEventCancel(): Boolean {
        setImageDrawable(mController.defaultIcon)
        setBackgroundResource(R.drawable.swipebtn_image_background_normal)
        return false
    }

    override fun onEventActionUp(whatNo: Int): Boolean {
        setImageDrawable(mController.defaultIcon)
        setBackgroundResource(R.drawable.swipebtn_image_background_normal)
        return false
    }

    override fun onEventActionDown(): Boolean {
        if (ThemeData.isEnabled() && ThemeData.getInstance().toolbarButtonBackgroundPress != null)
            background = ThemeData.getInstance().toolbarButtonBackgroundPress
        else
            setBackgroundResource(R.drawable.swipebtn_image_background_pressed)
        return false
    }

    override fun onChangeState(whatNo: Int) {
        setImageDrawable(mController.icon)
    }
}
